crt0stack/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+    ///
+    /// For more details see a [LWN article](https://lwn.net/Articles/631631/)
+    #[inline]
+    pub fn serialize(self, slice: &mut [u8]) -> Result<usize, Error> {
+        let mut sp = Stack::new(slice);
+        let mut str_ptr: [*const u8; SIZE] = [core::ptr::null(); SIZE];
+        let mut str_len: usize = 0;
+        let mut rand_ptr: Option<*const u8> = None;
+        let mut platform_ptr: Option<*const u8> = None;
+        let mut base_platform_ptr: Option<*const u8> = None;
+        let mut exec_fn_ptr: Option<*const u8> = None;
+
+        str_ptr[str_len] = core::ptr::null();
+        str_len += 1;
+
+        for s in self.envp.iter() {
+            if let Some(s) = s {
+                sp.push_l(0u8);
+                sp.push_slice(s.as_bytes());
+                str_ptr[str_len] = sp.ptr();
+                str_len += 1;
+            }
+        }
+
+        str_ptr[str_len] = core::ptr::null();
+        str_len += 1;
+
+        for s in self.argv.iter().rev() {
+            if let Some(s) = s {
+                sp.push_l(0u8);
+                sp.push_slice(s.as_bytes());
+                str_ptr[str_len] = sp.ptr();
+                str_len += 1;
+            }
+        }
+
+        sp.align_down(core::mem::size_of::<AuxvType>());
+
+        if let Some(x) = self.platform {
+            sp.push_l(0u8);
+            sp.push_slice(x.as_bytes());
+            platform_ptr = Some(sp.ptr());
+        }
+
+        if let Some(x) = self.base_platform {
+            sp.push_l(0u8);
+            sp.push_slice(x.as_bytes());
+            base_platform_ptr = Some(sp.ptr());
+        }
+
+        if let Some(x) = self.exec_fn {
+            sp.push_l(0u8);
+            sp.push_slice(x.as_bytes());
+            exec_fn_ptr = Some(sp.ptr());
+        }
+
+        if let Some(x) = self.random {
+            sp.align_down(core::mem::size_of::<AuxvType>());
+            sp.push_slice(&x);
+            rand_ptr = Some(sp.ptr());
+        }
+
+        sp.align_down(core::mem::size_of::<AuxvType>());
+
+        sp.push_l::<AuxvType>(0 as AuxvType);
+        sp.push_l::<AuxvKey>(AT_NULL);
+
+        match rand_ptr {
+            None => {}
+            Some(p) => {
+                sp.push_l::<AuxvType>(p as _);
+                sp.push_l::<AuxvKey>(AT_RANDOM);
+            }
+        }
+
+        match platform_ptr {
+            None => {}
+            Some(p) => {
+                sp.push_l::<AuxvType>(p as _);
+                sp.push_l::<AuxvKey>(AT_PLATFORM);
+            }
+        }
+
+        match base_platform_ptr {
+            None => {}
+            Some(p) => {
+                sp.push_l::<AuxvType>(p as _);
+                sp.push_l::<AuxvKey>(AT_BASE_PLATFORM);
+            }
+        }
+
+        match exec_fn_ptr {
+            None => {}
+            Some(p) => {
+                sp.push_l::<AuxvType>(p as _);
+                sp.push_l::<AuxvKey>(AT_EXECFN);
+            }
+        }
```

Comment:
Just thinking out loud: I wonder if we should be panicking if any of these fields are missing. I don't think it has to be super elaborate, we can just [unwrap](https://doc.rust-lang.org/core/option/enum.Option.html#method.unwrap) the options here as it will panic for us if they are `None`. Like this:

```diff
-        match rand_ptr {
-            None => {}
-            Some(p) => {
-                sp.push_l::<AuxvType>(p as _);
-                sp.push_l::<AuxvKey>(AT_RANDOM);
-            }
-        }
+        sp.push_l::<AuxvType>(rand_ptr.unwrap() as _);
+        sp.push_l::<AuxvKey>(AT_RANDOM);
```

What do you think?

If we decide to not just unwrap, then I think the following can be adapted to `if let` to shrink the code a little bit.

```diff
diff --git a/crt0stack/src/lib.rs b/crt0stack/src/lib.rs
index 15398ed..e324e56 100644
--- a/crt0stack/src/lib.rs
+++ b/crt0stack/src/lib.rs
@@ -301,37 +301,25 @@ impl<'a> Crt0Stack<'a> {
         sp.push_l::<AuxvType>(0 as AuxvType);
         sp.push_l::<AuxvKey>(AT_NULL);
 
-        match rand_ptr {
-            None => {}
-            Some(p) => {
-                sp.push_l::<AuxvType>(p as _);
-                sp.push_l::<AuxvKey>(AT_RANDOM);
-            }
-        }
+        if let Some(p) = rand_ptr {
+            sp.push_l::<AuxvType>(p as _);
+            sp.push_l::<AuxvKey>(AT_RANDOM);
+        };
 
-        match platform_ptr {
-            None => {}
-            Some(p) => {
-                sp.push_l::<AuxvType>(p as _);
-                sp.push_l::<AuxvKey>(AT_PLATFORM);
-            }
-        }
+        if let Some(p) = platform_ptr {
+            sp.push_l::<AuxvType>(p as _);
+            sp.push_l::<AuxvKey>(AT_PLATFORM);
+        };
 
-        match base_platform_ptr {
-            None => {}
-            Some(p) => {
-                sp.push_l::<AuxvType>(p as _);
-                sp.push_l::<AuxvKey>(AT_BASE_PLATFORM);
-            }
-        }
+        if let Some(p) = base_platform_ptr {
+            sp.push_l::<AuxvType>(p as _);
+            sp.push_l::<AuxvKey>(AT_BASE_PLATFORM);
+        };
 
-        match exec_fn_ptr {
-            None => {}
-            Some(p) => {
-                sp.push_l::<AuxvType>(p as _);
-                sp.push_l::<AuxvKey>(AT_EXECFN);
-            }
-        }
+        if let Some(p) = exec_fn_ptr {
+            sp.push_l::<AuxvType>(p as _);
+            sp.push_l::<AuxvKey>(AT_EXECFN);
+        };
 
         for s in self.auxv.iter() {
             if s.key == AT_NULL {
```

Reply:
panic!() should be appropriate